File: src/main/java/org/lanternpowered/server/network/entity/vanilla/RabbitEntityProtocol.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.entity.vanilla;

import org.lanternpowered.server.entity.LanternEntity;
import org.lanternpowered.server.network.entity.parameter.ParameterList;
import org.lanternpowered.server.registry.type.data.RabbitTypeRegistry;
import org.spongepowered.api.data.Keys;
import org.spongepowered.api.data.type.RabbitTypes;

public class RabbitEntityProtocol<E extends LanternEntity> extends AnimalEntityProtocol<E> {

    private int lastType;

    public RabbitEntityProtocol(E entity) {
        super(entity);
    }

    private int getTypeId() {
        return RabbitTypeRegistry.get().getId(this.entity.get(Keys.RABBIT_TYPE).orElseGet(RabbitTypes.WHITE));
    }

    @Override
    protected String getMobType() {
        return "minecraft:rabbit";
    }

    @Override
    protected void spawn(ParameterList parameterList) {
        super.spawn(parameterList);
        parameterList.add(EntityParameters.Rabbit.VARIANT, getTypeId());
    }

    @Override
    protected void update(ParameterList parameterList) {
        super.update(parameterList);
        final int type = this.getTypeId();
        if (type != this.lastType) {
            parameterList.add(EntityParameters.Rabbit.VARIANT, type);
            this.lastType = type;
        }
    }
}

File: src/main/kotlin/org/lanternpowered/server/network/entity/vanilla/RabbitEntityProtocol.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.entity.vanilla

import org.lanternpowered.api.data.Keys
import org.lanternpowered.api.key.NamespacedKey
import org.lanternpowered.api.key.minecraftKey
import org.lanternpowered.server.entity.LanternEntity
import org.lanternpowered.server.network.entity.parameter.ParameterList
import org.lanternpowered.server.registry.type.data.RabbitTypeRegistry
import org.spongepowered.api.data.type.RabbitTypes

class RabbitEntityProtocol<E : LanternEntity>(entity: E) : AnimalEntityProtocol<E>(entity) {

    companion object {
        private val TYPE = minecraftKey("rabbit")
    }

    private var lastTypeId = 0

    private val typeId: Int
        get() = RabbitTypeRegistry.getId(this.entity.get(Keys.RABBIT_TYPE).orElseGet(RabbitTypes.WHITE))

    override val mobType: NamespacedKey get() = TYPE

    override fun spawn(parameterList: ParameterList) {
        super.spawn(parameterList)
        parameterList.add(EntityParameters.Rabbit.VARIANT, this.typeId)
    }

    override fun update(parameterList: ParameterList) {
        super.update(parameterList)
        val type = this.typeId
        if (type != this.lastTypeId) {
            parameterList.add(EntityParameters.Rabbit.VARIANT, type)
            this.lastTypeId = type
        }
    }
}
